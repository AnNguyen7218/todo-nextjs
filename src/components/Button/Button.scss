@import 'config.gutters';
@import 'config.media';
@import 'config.colors';
@import 'config.fonts';

.button {
  display: inline-block;
  padding-top: gutter(8);
  padding-bottom: gutter(8);
  text-decoration: none;
  text-align: center;
  cursor: pointer;
  border: none;
  outline: none;

  @include break(xl) {
    padding-top: gutter(12);
    padding-bottom: gutter(12);
  }

  & > .button-children {
    display: flex;
    align-items: center;
    justify-content: center;
    text-align: center;
    text-decoration: none;
    color: $black;
    @include button;
    .loader {
      border-radius: 50%;
      border: 0.2rem solid $black;
      border-bottom-color: transparent;
      width: 1.8rem;
      height: 1.8rem;
      margin-right: 1.2rem;
      -webkit-animation: spin 2s linear infinite; /* Safari */
      animation: spin 2s linear infinite;
    }
  }

  &--fixed {
    padding-left: gutter(16);
    padding-right: gutter(16);

    @include break(xl) {
      padding-left: gutter(24);
      padding-right: gutter(24);
    }
  }

  &--scalable {
    display: inline-block;
    width: 100%;
    min-width: 12.5rem;
  }

  &--textOnly {
    background-color: unset;
    border: unset;

    &:hover {
      background-color: $text-only-press;
    }
    &:active {
      background-color: $text-only-press;
    }
  }

  &--border {
    border: 1px solid $black;
    background-color: transparent;

    &:hover {
      background-color: $text-only-press;
    }
    &:active {
      background-color: $text-only-press;
    }
  }

  &--containedColor {
    &.button--black {
      background-color: $black;
      &:hover {
        background-color: $contained-black;
      }
      &:active {
        background-color: $contained-black;
      }

      .button-children {
        color: $white;
        .loader {
          border-color: $white;
          border-bottom-color: transparent;
        }
      }
    }

    &.button--white {
      background-color: $white;
      &:hover {
        background-color: $contained-white;
      }
      &:active {
        background-color: $contained-white;
      }

      .button-children {
        color: $black;
        .loader {
          border-color: $black;
          border-bottom-color: transparent;
        }
      }
    }

    &.button--purple {
      background-color: $purple;
      &:hover {
        background-color: $contained-purple;
      }
      &:active {
        background-color: $contained-purple;
      }

      .button-children {
        color: $white;
        .loader {
          border-color: $white;
          border-bottom-color: transparent;
        }
      }
    }
  }

  &--disabled {
    pointer-events: none;
    cursor: not-allowed;
    opacity: 0.24;
  }
}

/* Safari */
@-webkit-keyframes spin {
  0% {
    -webkit-transform: rotate(0deg);
  }
  100% {
    -webkit-transform: rotate(360deg);
  }
}

@keyframes spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
