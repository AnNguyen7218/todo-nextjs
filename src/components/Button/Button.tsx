import React from 'react';
import Link from 'next/link';
import cx from 'classnames';
import css from './Button.scss';

type Props = {
  href?: string;
  as?: string;
  target?: string;
  disabled?: boolean;
  isBusy?: boolean;
  type?: 'button' | 'submit' | 'reset' | undefined;
  onClick?:
    | (((event?: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => void) &
        ((event?: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void))
    | undefined;
  children: React.ReactNode;
  id?: string;
  theme?: 'border' | 'textOnly' | 'containedColor';
  buttonType?: 'scalable' | 'fixed';
  color?: 'black' | 'white' | 'purple';
  className?: string;
  contentClass?: string;
};

type buttonComponentType = 'a' | 'button' | typeof Link;

const Button: React.FC<Props> = ({
  href = '',
  as = '',
  target,
  disabled,
  isBusy,
  type = 'button',
  onClick,
  children,
  id,
  className,
  contentClass,
  buttonType = 'fixed',
  theme = 'containedColor',
  color = 'black',
}) => {
  let ButtonComponent: buttonComponentType = 'button';
  if (href && (/^http/.test(href) || /^mailto/.test(href))) {
    ButtonComponent = 'a';
  } else if (href || as) {
    ButtonComponent = Link;
  } else if (target) {
    ButtonComponent = 'a';
  }

  const isLink = ButtonComponent === Link;
  const isButton = !isLink;

  const ButtonElement = isLink ? 'a' : 'button';

  const isDisabled = disabled || isBusy;

  const renderButton = () => {
    return (
      <ButtonElement
        disabled={isButton ? isDisabled : undefined}
        target={target}
        type={isButton ? type : undefined}
        onClick={isDisabled ? undefined : onClick}
        id={id}
        className={cx(
          css['button'],
          css[`button--${buttonType}`],
          css[`button--${theme}`],
          {
            [css[`button--${color}`]]: theme === 'containedColor',
            [css['button--disabled']]: disabled,
          },
          className
        )}
      >
        <div className={cx(css['button-children'], contentClass)}>
          {!isBusy ? (
            children
          ) : (
            <>
              <div className={css['loader']}></div>
              Please wait...
            </>
          )}
        </div>
      </ButtonElement>
    );
  };
  if (isLink) {
    return (
      <Link href={href} as={as}>
        {renderButton()}
      </Link>
    );
  }
  return renderButton();
};

export default Button;
