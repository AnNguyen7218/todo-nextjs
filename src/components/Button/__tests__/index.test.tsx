import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Button from '../Button';

afterEach(cleanup);

describe('Button Component', () => {
  test('should match button/link snapshot', () => {
    const { container } = render(
      <>
        <Button href="/about">Button Component</Button>
        <Button href="https://google.com">Button Component</Button>
        <Button>Button Component</Button>
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('should match button type', () => {
    const { container } = render(
      <>
        <Button href="/about" buttonType="fixed">
          Button Component
        </Button>
        <Button href="https://google.com" buttonType="fixed">
          Button Component
        </Button>
        <Button buttonType="fixed">Button Component</Button>
        <Button href="/about" buttonType="scalable">
          Button Component
        </Button>
        <Button href="https://google.com" buttonType="scalable">
          Button Component
        </Button>
        <Button buttonType="scalable">Button Component</Button>
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('should match button style', () => {
    const { container } = render(
      <>
        <Button theme="textOnly">Button Component</Button>
        <Button theme="border">Button Component</Button>
        <Button theme="containedColor">Button Component</Button>
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('should match button contained color style', () => {
    const { container } = render(
      <>
        <Button theme="containedColor" color="purple">
          Button Component
        </Button>
        <Button theme="containedColor" color="black">
          Button Component
        </Button>
        <Button theme="containedColor" color="white">
          Button Component
        </Button>
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('should match button disable style', () => {
    const { container } = render(
      <>
        <Button theme="textOnly" disabled>
          Button Component
        </Button>
        <Button theme="border" disabled>
          Button Component
        </Button>
        <Button theme="containedColor" color="purple" disabled>
          Button Component
        </Button>
        <Button theme="containedColor" color="black" disabled>
          Button Component
        </Button>
        <Button theme="containedColor" color="white" disabled>
          Button Component
        </Button>
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('should match button busy style', () => {
    const { container } = render(
      <>
        <Button href="/about" theme="textOnly" isBusy>
          Button Component
        </Button>
        <Button theme="border" isBusy>
          Button Component
        </Button>
        <Button theme="containedColor" color="purple" isBusy>
          Button Component
        </Button>
        <Button theme="containedColor" color="black" isBusy>
          Button Component
        </Button>
        <Button theme="containedColor" color="white" isBusy>
          Button Component
        </Button>
      </>
    );
    expect(container).toMatchSnapshot();
  });
});
