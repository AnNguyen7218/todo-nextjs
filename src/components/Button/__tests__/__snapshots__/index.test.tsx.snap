// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should match button busy style 1`] = `
<div>
  <a
    class="button button--fixed button--textOnly"
    href="/about"
  >
    <div
      class="button-children"
    >
      <div
        class="loader"
      />
      Please wait...
    </div>
  </a>
  <button
    class="button button--fixed button--border"
    disabled=""
    type="button"
  >
    <div
      class="button-children"
    >
      <div
        class="loader"
      />
      Please wait...
    </div>
  </button>
  <button
    class="button button--fixed button--containedColor button--purple"
    disabled=""
    type="button"
  >
    <div
      class="button-children"
    >
      <div
        class="loader"
      />
      Please wait...
    </div>
  </button>
  <button
    class="button button--fixed button--containedColor button--black"
    disabled=""
    type="button"
  >
    <div
      class="button-children"
    >
      <div
        class="loader"
      />
      Please wait...
    </div>
  </button>
  <button
    class="button button--fixed button--containedColor button--white"
    disabled=""
    type="button"
  >
    <div
      class="button-children"
    >
      <div
        class="loader"
      />
      Please wait...
    </div>
  </button>
</div>
`;

exports[`Button Component should match button contained color style 1`] = `
<div>
  <button
    class="button button--fixed button--containedColor button--purple"
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
  <button
    class="button button--fixed button--containedColor button--black"
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
  <button
    class="button button--fixed button--containedColor button--white"
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
</div>
`;

exports[`Button Component should match button disable style 1`] = `
<div>
  <button
    class="button button--fixed button--textOnly button--disabled"
    disabled=""
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
  <button
    class="button button--fixed button--border button--disabled"
    disabled=""
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
  <button
    class="button button--fixed button--containedColor button--purple button--disabled"
    disabled=""
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
  <button
    class="button button--fixed button--containedColor button--black button--disabled"
    disabled=""
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
  <button
    class="button button--fixed button--containedColor button--white button--disabled"
    disabled=""
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
</div>
`;

exports[`Button Component should match button style 1`] = `
<div>
  <button
    class="button button--fixed button--textOnly"
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
  <button
    class="button button--fixed button--border"
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
  <button
    class="button button--fixed button--containedColor button--black"
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
</div>
`;

exports[`Button Component should match button type 1`] = `
<div>
  <a
    class="button button--fixed button--containedColor button--black"
    href="/about"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </a>
  <button
    class="button button--fixed button--containedColor button--black"
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
  <button
    class="button button--fixed button--containedColor button--black"
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
  <a
    class="button button--scalable button--containedColor button--black"
    href="/about"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </a>
  <button
    class="button button--scalable button--containedColor button--black"
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
  <button
    class="button button--scalable button--containedColor button--black"
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
</div>
`;

exports[`Button Component should match button/link snapshot 1`] = `
<div>
  <a
    class="button button--fixed button--containedColor button--black"
    href="/about"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </a>
  <button
    class="button button--fixed button--containedColor button--black"
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
  <button
    class="button button--fixed button--containedColor button--black"
    type="button"
  >
    <div
      class="button-children"
    >
      Button Component
    </div>
  </button>
</div>
`;
